RevDiff: Add hotkeys to stage/unstage/reset

gerhardol opened the pull request:
Separating hotkey handling in RevDiff file status list from #7825 

No longer based on #8193 (some conflicts)

## Proposed changes

Add hotkeys to stage/unstage/reset in RevDiff
Checks that the status list is in focus (if the viewer is in focus in #7825, lines will be changed instead).


## Screenshots <!-- Remove this section if PR does not change UI -->

### After

![image](https://user-images.githubusercontent.com/6248932/84868761-79422500-b07d-11ea-96e6-89fdd21b5335.png)

![image](https://user-images.githubusercontent.com/6248932/84868884-a0005b80-b07d-11ea-862c-7ed46c4017e1.png)

![image](https://user-images.githubusercontent.com/6248932/85204808-503cc100-b317-11ea-84a3-67be2df429a7.png)


## Test methodology 

Manual

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

mstv commented:
> Note that there is no "are you sure?" for the reset

You're right. This is a dangerous function in combination with the hotkey `R`, which is prone to be hit by mistake. (We even had an issue complaining about the single-character hotkeys in the FileStatusList. I strongly vote for keeping them anyways.)
So we should add the confirmation popup for the hotkey here, too.
For the context menu item, it _could_ be omitted because the user must navigate to the submenu.

RussKie commented:
We have to be really careful here because we will be breaking an establish
user behavior.

If we introduce a warning message it has to be permanently dismissable (use
TaskDialog with Verification).


On Fri, Jun 19, 2020, 7:58 AM Michael Seibt <notifications@github.com>
wrote:

> Note that there is no "are you sure?" for the reset
>
> You're right. This is a dangerous function in combination with the hotkey
> R, which is prone to be hit by mistake. (We even had an issue complaining
> about the single-character hotkeys in the FileStatusList. I strongly vote
> for keeping them anyways.)
> So we should add the confirmation popup for the hotkey here, too.
> For the context menu item, it *could* be omitted because the user must
> navigate to the submenu.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/8239#issuecomment-646326379>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXVAIAPLM3V5NPQRE33RXKEYZANCNFSM4OAKKZSA>
> .
>

gerhardol commented:
Rebased on master, to speedup reviews

Added a message for the hotkey invocation only, similar to other "Reset" invocations. The menu indication here is the only situation I see that do not open a dialog.

The message here is a little special as I want to include the commit to revert to (this would not be needed for invocation from the menu).

>We have to be really careful here because we will be breaking an establish
user behavior.
If we introduce a warning message it has to be permanently dismissable (use
TaskDialog with Verification).

Change defaults is dangerous.
A dismissible dialog is a separate feature, should be global (we do not want to add 10 new options).

mstv commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -948,5 +971,62 @@ public void SwitchFocus(bool alreadyContainedFocus)
+        /// <returns>true if executed</returns>
+        private bool StageSelectedFiles()
+        {
+            var selectedIds = DiffFiles.SelectedItems.SecondIds().ToList();
+            if (!DiffFiles.Focused || selectedIds.Count != 1 || selectedIds.FirstOrDefault() != ObjectId.WorkTreeId)
+            {
+                return false;
+            }
```
The meaning of the return value is rather "hotkey handled" than "command executed".
IIUIC, the return value is used to find another control which handles the same hotkey.
I would not support the reuse of these hotkeys.
So this check should be omitted and the function always returns true. Then we do not need a dedicated function.

gerhardol replied:
The check still need to be there, but return true (since early return is preferred rather than check if allowed).

mstv replied:
Ah OK. Then I would return nothing, i.e. void, and return true directly in `ExecuteCommand`.

Why must `DiffFiles` be focused?
The hotkeys will be reused for operations on selected lines in the `DiffText`.
I think it suffices to check whether `DiffText` has the focus. Then "selected lines" else "selected files".

gerhardol replied:
The logic here (with #7825 included) is the same as in FormCommit, so I propose to keep it and check for FileStatusList focus. 
In #7825 the method return false if it could not be executed, not sure what should be done here...
But I will change these functions to void as suggested.

gerhardol replied:
Changed so return false if FileStatusList is not focused only and consider the key consumed otherwise
Other keyhandling may cause both hotkey and script execution, should probably be changed.

mstv commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -948,5 +971,62 @@ public void SwitchFocus(bool alreadyContainedFocus)
+        /// <returns>true if executed</returns>
+        private bool UnstageSelectedFiles()
+        {
+            var selectedIds = DiffFiles.SelectedItems.SecondIds().ToList();
+            if (!DiffFiles.Focused || selectedIds.Count != 1 || selectedIds.FirstOrDefault() != ObjectId.IndexId)
+            {
+                return false;
+            }
```
ditto

mstv commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -948,5 +971,62 @@ public void SwitchFocus(bool alreadyContainedFocus)
+        /// <returns>true if executed</returns>
+        private bool ResetSelectedFiles()
+        {
+            var parentIds = DiffFiles.SelectedItems.FirstIds().ToList();
+            if (!DiffFiles.Focused || parentIds.Count != 1 || !CanResetToRevision(parentIds.FirstOrDefault()))
+            {
+                return false;
+            }
```
ditto

mstv commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -948,5 +971,62 @@ public void SwitchFocus(bool alreadyContainedFocus)
+        /// Hotkey handler
+        /// </summary>
+        /// <returns>true if executed</returns>
+        private bool ResetSelectedFiles()
```
```suggestion
        private void ResetSelectedFilesWithConfirmation()
```

mstv commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -948,5 +971,62 @@ public void SwitchFocus(bool alreadyContainedFocus)
+            }
+
+            var rev = _firstRevision.Text + (DescribeRevision(DiffFiles.SelectedItems.FirstRevs().ToList()) ?? string.Empty);
+            var text = string.Format(_resetSelectedChangesText.Text, rev);
+            if (MessageBox.Show(this, text, _resetChangesCaption.Text, MessageBoxButtons.YesNo, MessageBoxIcon.Question) !=
+                DialogResult.Yes)
```
Move to `MessageBoxes`, please.

gerhardol replied:
ok

mstv commented on GitUI/MessageBoxes.cs:
```diff
@@ -118,5 +119,11 @@ public static void ShellNotFound([CanBeNull] IWin32Window owner)
             MessageBox.Show(owner, Instance._shellNotFound.Text, Instance._shellNotFoundCaption.Text, MessageBoxButtons.OK, MessageBoxIcon.Error);
         }
+
+        public static bool ResetSelectedFiles(IWin32Window owner, string text)
+        {
+            return MessageBox.Show(owner, text, Instance._resetChangesCaption.Text, MessageBoxButtons.YesNo, MessageBoxIcon.Question) ==
+                   DialogResult.Yes;
+        }
```
```suggestion
        public static bool ResetSelectedFiles(IWin32Window owner, string text)
            => MessageBox.Show(owner, text, Instance._resetChangesCaption.Text, MessageBoxButtons.YesNo, MessageBoxIcon.Question) == DialogResult.Yes;
```

gerhardol replied:
will squash at approval

mstv commented:
@msftbot merge in 3 days
